### ISSUE TITLE
Variables cannot be used in LocationAt

### ISSUE THREAD

**Issue by iOshawott** [SkriptLang/Skript#5827]
### Skript/Server Version

```markdown
[20:36:14 INFO]: [Skript] Skript's aliases can be found here: https://github.com/SkriptLang/skript-aliases
[20:36:14 INFO]: [Skript] Skript's documentation can be found here: https://docs.skriptlang.org/
[20:36:14 INFO]: [Skript] Skript's tutorials can be found here: https://docs.skriptlang.org/tutorials
[20:36:14 INFO]: [Skript] Server Version: git-Paper-72 (MC: 1.20.1)
[20:36:14 INFO]: [Skript] Skript Version: 2.7.0-beta3 (skriptlang-github)
[20:36:14 INFO]: [Skript] Installed Skript Addons: None
[20:36:14 INFO]: [Skript] Installed dependencies: 
[20:36:14 INFO]: [Skript]  - Vault v1.7.3-b131
[20:36:14 INFO]: [Skript]  - WorldGuard v7.0.9-SNAPSHOT+2242-c4295d7
```

### Bug Description

While trying to set location using variables I'm getting error "Variables cannot be used in location expression". **It was fully working in skript 2.6.3** so I'm posting it to be sure - is it bug or expected behavior? If expected - is there any workaround I can use to keep that functionality known from 2.6.3?

code:
set {_cord2} to the location at {_x2}, 0, {_z1} of the world world

[18:55:31 INFO]: [Skript] Line 40: (cuboid.sk)
[18:55:31 INFO]:     Variables cannot be used here.
[18:55:31 INFO]:     Line: set {_cord2} to the location at {_x2}, 0, {_z1} of the world world

### Expected Behavior

Expected behavior is not to getting error Variables cannot be used here.

### Steps to Reproduce

Try to set variable to location using variables (ex. set {_cord2} to the location at {_x2}, 0, {_z1} of the world world)

### Agreement

- [X] I have read the guidelines above and affirm I am following them with this report.

**Comment by TheLimeGlass** [comment]
Can you show how you're setting those variables `{_x2} and {_z1}`

**Comment by iOshawott** [comment]
for example, that simple code was working on 2.6:
`command /testcub:
	trigger:
		set {_x1} to 5
		set {_z1} to 10
		set {ncs1::%player%} to the location at {_x1}, 0, {_z1} of the world world`

but gives error on 2.7

**Comment by TheLimeGlass** [comment]
This error is present in 2.6.4 not 2.6.3

**Comment by TheLimeGlass** [comment]
Issue is caused by the `world world`

You can use the following to solve your issue until fixed.
```vb
location at {_x1}, 0, {_z1} of the world "world"

or

location({_x1}, 0, {_z1}, world "world")
```

**Comment by Mr-Darth** [comment]
Caused by #5171. `world world` tries to do `world (world of <default expr>)` (the first `world` is part of ExprLocationAt's pattern). The aforementioned PR changed the way event value lookup works: it does [not search in superclasses](https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/registrations/EventValues.java#L313) anymore.

In this case, there is a [`CommandSender` event value registered for `CommandEvent`](https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java#L1002), but the actual event of the trigger is `ScriptCommandEvent`. Since `!ScriptCommandEvent.class.isAssignableFrom(CommandEvent.class)` is true, this value is skipped, so Skript doesn't have anything to get a world from.

PR #5839 is breaking and does not fix the underlying issue.

**Comment by AyhamAl-Ali** [comment]
> Caused by #5171. `world world` tries to do `world (world of <default expr>)` (the first `world` is part of ExprLocationAt's pattern). The aforementioned PR changed the way event value lookup works: it does [not search in superclasses](https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/registrations/EventValues.java#L313) anymore.
> 
> In this case, there is a [`CommandSender` event value registered for `CommandEvent`](https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java#L1002), but the actual event of the trigger is `ScriptCommandEvent`. Since `!ScriptCommandEvent.class.isAssignableFrom(CommandEvent.class)` is true, this value is skipped, so Skript doesn't have anything to get a world from.
> 
> PR #5839 is breaking and does not fix the underlying issue.

Good job on tracing this and good explanation 👌 I think this should just be marked as invalid and the correct usage of world syntax is `world "world"`.

**Comment by TheLimeGlass** [comment]
> Caused by #5171. `world world` tries to do `world (world of <default expr>)` (the first `world` is part of ExprLocationAt's pattern). The aforementioned PR changed the way event value lookup works: it does [not search in superclasses](https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/registrations/EventValues.java#L313) anymore.
> 
> In this case, there is a [`CommandSender` event value registered for `CommandEvent`](https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java#L1002), but the actual event of the trigger is `ScriptCommandEvent`. Since `!ScriptCommandEvent.class.isAssignableFrom(CommandEvent.class)` is true, this value is skipped, so Skript doesn't have anything to get a world from.
> 
> PR #5839 is breaking and does not fix the underlying issue.

I already tried that, reverting that exact pull request did not change anything.

**Comment by Mr-Darth** [comment]
> > Caused by #5171. `world world` tries to do `world (world of <default expr>)` (the first `world` is part of ExprLocationAt's pattern). The aforementioned PR changed the way event value lookup works: it does [not search in superclasses](https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/registrations/EventValues.java#L313) anymore.
> > 
> > In this case, there is a [`CommandSender` event value registered for `CommandEvent`](https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java#L1002), but the actual event of the trigger is `ScriptCommandEvent`. Since `!ScriptCommandEvent.class.isAssignableFrom(CommandEvent.class)` is true, this value is skipped, so Skript doesn't have anything to get a world from.
> > 
> > PR #5839 is breaking and does not fix the underlying issue.
> 
> I already tried that, reverting that exact pull request did not change anything. This is not correct.
> 
> The pull request to fix this, does solve the issue, but as Ali pointed out, it cannot be used in literal context like a world event structure. I'll resolve that eventually.

I, too, have already tried that, and reverting the exact pull request did fix it. This is correct.

Yes, your pull request does (kind of) solve the issue, but it's more of a band-aid fix than an actual solution. First of all, the code would've previously grabbed the world of the sender's location, but with your PR, it will grab the "world" world. Plus, not only does it break world literals, but it also allows random nonsense like `broadcast bananas` if a world with that name exists.

The problem here is that converting event values does not work from subclass to superclass. There's no problem with ExprWorldFromName having the same pattern because classinfo literals are always tried last.

BTW, a better workaround would be to use `event-world` instead.

**Comment by UnderscoreTud** [comment]
@TheLimeGlass You are wrong. I have reverted your PR #5171 and the code provided in the example works perfectly fine now.
```hs
command /hello:
    trigger:
        set {_cord2} to the location at {_x2}, 0, {_z1} of the world world
```
![image](https://github.com/SkriptLang/Skript/assets/98935832/ccae4c47-7070-4acf-93a7-31477df98ff6)
[Here's the branch](https://github.com/UnderscoreTud/Skript/tree/fix/revert-pr)

**Comment by TheLimeGlass** [comment]
> @TheLimeGlass You are wrong. I have reverted your PR #5171 and the code provided in the example works perfectly fine now.
> 
> ```haskell
> command /hello:
>     trigger:
>         set {_cord2} to the location at {_x2}, 0, {_z1} of the world world
> ```
> 
> ![image](https://user-images.githubusercontent.com/98935832/258666424-ccae4c47-7070-4acf-93a7-31477df98ff6.png) [Here's the branch](https://github.com/UnderscoreTud/Skript/tree/fix/revert-pr)

https://github.com/SkriptLang/Skript/pull/5171 did not touch EventValueExpression nor getExactEventValueGetter. You're completely removing functionality not related to the original pull request.
